torch/distributed/_checkpoint/storage_reader.py

```diff
@@ -0,0 +1,96 @@
+import abc
+import operator
+import os
+import pickle
+from typing import List, Optional, cast
+
+import torch
+from torch import Tensor
+from torch.futures import Future
+
+from .metadata import BytesReadRequest, Metadata, TensorReadRequest
+
+
+class StorageReader(abc.ABC):
+    """
+    Interface to read from the underlying storage system.
+    """
+
+    @abc.abstractmethod
+    def read_bytes(self, requests: List[BytesReadRequest]) -> Future[None]:
+        """
+        Read request and returns a Future to wait on.
+        Args:
+            requests (List[BytesReadRequest]): see `./metadata.py`]
+        """
+        pass
+
+    @abc.abstractmethod
+    def read_tensors(self, requests: List[TensorReadRequest]) -> Future[None]:
+        """
+        Performs a read request and returns a Future to wait on.
+        Args:
+            requests (List[BytesReadRequest]): see `./metadata.py`
+        """
+        pass
+
+    @abc.abstractmethod
+    def read_metadata(self) -> Metadata:
+        """
+        Read the meta data and returns.
+        """
+        pass
+
+
+class FileSystemReader(StorageReader):
+    def __init__(self, path: str) -> None:
+        super().__init__()
+        self.path = path
+
+    def read_tensors(self, requests: List[TensorReadRequest]) -> Future[None]:
+        """
+        Very basic implementation that read from file system.
+        """
+        # Sort the the requests by storage key and try to reuse the loaded tensors
+        requests.sort(key=operator.attrgetter("storage_key"))
+
+        cached_storage_key = None
+        view_cached: Optional[Tensor] = None
+
+        for req in requests:
+            if cached_storage_key != req.storage_key:
+                with open(os.path.join(self.path, req.storage_key), "rb") as storage:
+                    view_cached = cast(Tensor, torch.load(storage))
+                    cached_storage_key = req.storage_key
+
+            view_to_copy: Tensor = cast(Tensor, view_cached)
+            # FileSystemWrite writes the tensor as is during save.
+            # During load time, we will load the Tensor (with it orignal view)
+            # narrow it along all dimemsions, and copy_ it to the
+            # target tensor, which will be the same size.
+            for dim, (start, length) in enumerate(zip(req.offsets, req.lengths)):
+                view_to_copy = torch.narrow(view_to_copy, dim, start, length)
```

Comment:
This issue is not as serious as it looks since we only keep one tensor in memory at a time.

We decided in favor of this tradeoff as it's enables us to support arbitrary tensors while the alternative would require to effectively reimplement a large part of torch.load / torch.save.

Tracking it here: https://github.com/pytorch/pytorch/issues/76331